xwakes/wit/component.py

```diff
@@ -385,3 +510,896 @@ def get_shorthand_type(self) -> str:
+                interpolation_frequencies: ArrayLike = None,
+                impedance_input: Optional[Callable] = None,
+                interpolation_times: ArrayLike = None,
+                wake_input: Optional[Callable] = None,
+                kind: str = None,
+                plane: str = None,
+                source_exponents: Tuple[int, int] = None,
+                test_exponents: Tuple[int, int] = None,
+                name: str = "Interpolated Component",
+                f_rois: Optional[List[Tuple[float, float]]] = None,
+                t_rois: Optional[List[Tuple[float, float]]] = None):
+
+        assert ((interpolation_frequencies is not None) ==
+                (impedance_input is not None)), ("Either both or none of the "
+                "impedance and the interpolation frequencies must be given")
+
+        source_exponents, test_exponents, plane = _handle_plane_and_exponents_input(
+                            kind=kind, exponents=None,
+                            source_exponents=source_exponents,
+                            test_exponents=test_exponents,
+                            plane=plane)
+
+        self.interpolation_frequencies = interpolation_frequencies
+        self.impedance_samples = impedance_input(self.interpolation_frequencies)
+
+        assert ((interpolation_times is not None) ==
+                (wake_input is not None)), ("Either both or none of the wake "
+                "and the interpolation times must be given")
+
+        self.interpolation_times = interpolation_times
+        self.wake_samples = wake_input(self.interpolation_times)
+
+        super().__init__(impedance=lambda x: 0, wake=lambda x: 0, plane=plane,
+                       source_exponents=source_exponents,
+                       test_exponents=test_exponents,
+                       f_rois=f_rois, t_rois=t_rois,
+                       name=name)
+
+    def impedance(self, f):
+        if hasattr(self, 'impedance_samples') is not None:
+            return np.interp(f, self.interpolation_frequencies,
+                             self.impedance_samples)
+        else:
+            return np.zeros_like(f)
+
+    def wake(self, t):
+        if hasattr(self, 'wake_samples'):
+            return np.interp(t, self.interpolation_times, self.wake_samples,
+                             left=0, right=0 # pad with zeros outside the range
+            )
+        else:
+            return np.zeros_like(t)
+
+    def function_vs_t(self, t, beta0, dt):
+
+        assert dt > 0
+
+        isscalar = np.isscalar(t)
+        if isscalar:
+            t = np.array([t])
+
+        out = self.wake(t)
+
+        # Handle discontinuity at edges (consistently with beam loading theorem)
+        mask_left_edge = np.abs(t - self.interpolation_times[0]) < dt / 2
+        out[mask_left_edge] = self.wake_input[0] / 2.
+        mask_right_edge = np.abs(t - self.interpolation_times[-1]) < dt / 2
+        out[mask_right_edge] = self.wake_input[-1] / 2.
```

Comment:
Yes but I am also confused because since we know the sign of t and interpolation_times (which is assumed positive by convention) we should be able to get rid of one of the two ..  I need to investigate.

Reply:
In the end the code here is fine. A comment was added to explain it better. Thanks @giadarol !